peak/mapper/binding.py

```diff
@@ -0,0 +1,195 @@
+        enumeration_scheme : tp.Mapping[type,tp.Callable] = {}
+    ):
+        self.smt_cache = SMT_cache()
+        self.enumeration_scheme = enumeration_scheme
+        assert issubclass(arch_isa,Product)
+        assert issubclass(ir_isa,Product)
+        self.arch_isa = arch_isa
+        self.ir_isa = ir_isa
+
+        self.arch_flat = _flatten_adt(self.arch_isa)
+        self.ir_flat = _flatten_adt(self.ir_isa)
+
+        arch_by_t = _sort_by_t(self.arch_flat)
+        ir_by_t = _sort_by_t(self.ir_flat)
+
+        def _has_binding(arch_by_t,ir_by_t):
+            #for each type, each input of the type in the IR needs to at least be able to bind to one other in the arch
+            for t in ir_by_t:
+                if not (t in arch_by_t):
+                    return False
+                if len(arch_by_t[t]) < len(ir_by_t[t]):
+                    return False
+            return True
+
+        #Check if there is at least one input binding
+        self.has_binding = _has_binding(arch_by_t,ir_by_t)
+
+        #check for early out
+        if not self.has_binding:
+            return
+        possible_matching = {}
+        for arch_type, arch_paths in arch_by_t.items():
+            if is_adt_type(arch_type): #Sum or Enum
+                unbound_possibilities = ("E",)
+            elif allow_exists:
+                unbound_possibilities = ("A","E")
+            else:
+                unbound_possibilities = ("A",)
+
+            #Returns this list of things that match type t from arch
+            ir_paths = ir_by_t.setdefault(arch_type, [])
+            num_unbound = len(arch_paths) - len(ir_paths)
+
+            #Create a potentials list (to be passed to product)
+            #This will tie each unbound variable with either "A" or "E"
+            ir_path_potentials = list(it.chain(((path,) for path in ir_paths), it.repeat(unbound_possibilities,num_unbound)))
+            assert len(ir_path_potentials) == len(arch_paths)
+            for ir_path in it.product(*ir_path_potentials):
+                #For every permutation of arch, match it with ir
+                for arch_perm in it.permutations(arch_paths):
+                    possible_matching.setdefault(arch_type, []).append(list(zip(ir_path, arch_perm)))
+
+        self.possible_matching = possible_matching
+
+        del arch_by_t
+        del ir_by_t
+
+    #This will yield a "binding" which is a list of (ir_path,arch_path) pairs
+    def enumerate(self):
+        assert self.has_binding
+        for l in it.product(*self.possible_matching.values()):
+            yield it.chain(*l)
+
+    def get_enumerate(self,t):
+        #custom enumeration
+        if t in self.enumeration_scheme:
+            return self.enumeration_scheme[t]
+
+        #default scheme
+        if is_adt_type(t):
+            def gen():
+                return t.enumerate()
+            return gen
+        elif issubclass(t,SMTBitVector):
+            def gen():
```

Comment:
should be setdefault in \_\_init\_\_ but there is the issue of the each size needs to be registered.